### PR TITLE
Implement order hashing, add asset data classes, add tests

### PR DESCRIPTION
**Pull request by bojan96**
Title says it

### CODE REVIEW

File: 0x/Assets/ERC20Asset.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+namespace ZeroX.Assets
+{
+
+    /// <summary>
+    /// ERC20 asset
+    /// </summary>
+    public class ERC20Asset : TokenAsset
+    {
+
+        private static readonly byte[] _erc20AssetHeader = new byte[] { 0xf4, 0x72, 0x61, 0xb0 };
```
Comment (gpufreak) [inline 0x/Assets/ERC20Asset.cs]: I would put this in constants file.

File: 0x/Assets/ERC721Asset.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+namespace ZeroX.Assets
+{
+
+    /// <summary>
+    /// ERC721 Asset
+    /// </summary>
+    public class ERC721Asset : TokenAsset
+    {
+
+        private static readonly byte[] _erc721AssetHeader = new byte[] { 0x02, 0x57, 0x17, 0x92 };
```
Comment (gpufreak) [inline 0x/Assets/ERC721Asset.cs]: Also, I would put this in some contants file.

File: 0x/Orders/Order.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        public EthereumAddress MakerAddress { get; set; }
+        public EthereumAddress TakerAddress { get; set; }
+        public EthereumAddress FeeRecipientAddress { get; set; }
+        public EthereumAddress SenderAddress { get; set; }
+        public BigInteger MakerAssetAmount { get; set; }
+        public BigInteger TakerAssetAmount { get; set; }
+        public BigInteger MakerFee { get; set; }
+        public BigInteger TakerFee { get; set; }
+        public BigInteger ExpirationTimeSeconds { get; set; }
+        public BigInteger Salt { get; set; }
```
Comment (gpufreak) [inline 0x/Orders/Order.cs]: Is this necessary a number?

Comment (gpufreak) [reply 0x/Orders/Order.cs]: Salt*

File: 0x/Orders/Order.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        /// <exception cref="ArgumentNullException">exchangeAddress is null</exception>
+        public byte[] Hash(EthereumAddress exchangeAddress)
+        {
+            if (exchangeAddress == null)
+                throw new ArgumentNullException(nameof(exchangeAddress));
+
+            EIP712Domain domain = new EIP712Domain()
+            {
+                Name = "0x Protocol",
+                Version = "2",
```
Comment (gpufreak) [inline 0x/Orders/Order.cs]: Maybe the version can be a property that's changable afterwards.

File: 0x/Orders/Order.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                VerifyingContract = exchangeAddress.ToString()
+            };
+
+            return EIP712.EIP712.Hash(EIP712Order, domain);
```
Comment (gpufreak) [inline 0x/Orders/Order.cs]: `EIP712.EIP712.Hash()` looks a bit weird.

File: 0x/Utilities/EthereumAddress.cs [hunk 6]
```diff
@@ -7,13 +7,17 @@ namespace ZeroX.Utilities
 {
     public class EthereumAddress
     {
+
+        // TODO: Add docs
+
         private string _address;
         private const string ZeroAddressLiteral = "0x0000000000000000000000000000000000000000";
```
Comment (gpufreak) [inline 0x/Utilities/EthereumAddress.cs]: Also in a constants file (reuse if you have it in some dependency).

File: 0x/Utilities/EthereumAddress.cs [hunk 7]
```diff
@@ -57,5 +61,13 @@ public override string ToString()
                 throw new InvalidCastException($"Can not cast string to EthereumAddress", ex);
```
Comment (gpufreak) [inline 0x/Utilities/EthereumAddress.cs]: `InvalidCastException` or `ArgumentNullException`?

File: Tests/Tests/AssetDataTests.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+{
+    [TestClass]
+    public class AssetDataTests
+    {
+
+        private static readonly EthereumAddress _testAddress = (EthereumAddress)Constants.TestEthereumAddress;
+        private static readonly byte[] _testERC20AssetData = ("0xf47261b0000000000000000000000000646934" +
+                "251e9045f41ddf7145dea5cd63cd45d8da").HexToByteArray();
+
+        private static readonly byte[] _testERC721AssetData = ("0x0257179200000000000000000000000064693425" +
```
Comment (gpufreak) [inline Tests/Tests/AssetDataTests.cs]: How much example coverage do you have? If it's only covering like 10% of the input data, then you can make an array of examples that you feed into every test (like input arguments).

File: Tests/Tests/Constants.cs [hunk 9]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Tests.Tests
+{
+    public static class Constants
+    {
+        public const string TestEthereumAddress = "0x646934251e9045F41DdF7145Dea5cD63cD45d8Da";
```
Comment (gpufreak) [inline Tests/Tests/Constants.cs]: This whole class can contain all 0x related addresses (on the popular Ethereum blockchains that we use).

File: 0x/Assets/ERC20Asset.cs (hunk 1 above)
Comment (bojan96) [reply 0x/Assets/ERC20Asset.cs]: I would not, currently this is only used in this class.

File: 0x/Assets/ERC721Asset.cs (hunk 2 above)
Comment (bojan96) [reply 0x/Assets/ERC721Asset.cs]: I would not, currently this is only used in this class.

File: 0x/Orders/Order.cs (hunk 3 above)
Comment (bojan96) [reply 0x/Orders/Order.cs]: Salt is BigNumber because corresponding member in solidity struct is of type "uint256"

File: 0x/Orders/Order.cs (hunk 5 above)
Comment (bojan96) [reply 0x/Orders/Order.cs]: That is why asked how to name EIP712 class, referring to EIP712 in code resolves to namespace(using EIP712 does not help)

File: 0x/Utilities/EthereumAddress.cs (hunk 7 above)
Comment (bojan96) [reply 0x/Utilities/EthereumAddress.cs]: This is constructor so ArgumentNullException is suitable.

File: Tests/Tests/Constants.cs (hunk 9 above)
Comment (bojan96) [reply Tests/Tests/Constants.cs]: Yes

File: Tests/Tests/AssetDataTests.cs (hunk 8 above)
Comment (bojan96) [reply Tests/Tests/AssetDataTests.cs]: I will try to add more data in future PRs

File: 0x/Utilities/EthereumAddress.cs (hunk 6 above)
Comment (bojan96) [reply 0x/Utilities/EthereumAddress.cs]: Same as above

**Pull request merged**